packages/grid/src/vaadin-grid.js

```diff
@@ -535,42 +535,73 @@ class Grid extends ElementMixin(
     }
   }
 
+  __getIntrinsicWidth(col) {
+    const initialWidth = col.width;
+    const initialFlexGrow = col.flexGrow;
+
+    col.width = 'auto';
+    col.flexGrow = 0;
+
+    let width = 0;
+
+    // Note: _allCells only contains cells which are currently rendered in DOM
+    col._allCells
+      .filter((c) => {
+        // Exclude body cells that are out of the visible viewport
+        return !this.$.items.contains(c) || this._isInViewport(c.parentElement);
+      })
+      .forEach((c) => {
+        // Add 1px buffer to the offset width to avoid too narrow columns (sub-pixel rendering)
+        const cellWidth = c.offsetWidth + 1;
+        width = Math.max(width, cellWidth);
+      });
```

Comment:
Minor cleanup

```suggestion
    // Note: _allCells only contains cells which are currently rendered in DOM
    const width = col._allCells
      .filter((cell) => {
        // Exclude body cells that are out of the visible viewport
        return !this.$.items.contains(cell) || this._isInViewport(cell.parentElement);
      })
      .reduce((width, cell) => {
        // Add 1px buffer to the offset width to avoid too narrow columns (sub-pixel rendering)
        return Math.max(width, cell.offsetWidth + 1);
      }, 0);
```